optuna/integration/botorch.py

```diff
@@ -0,0 +1,507 @@
+    configurations. Parameters are transformed to continuous space and passed to BoTorch, and then
+    transformed back to Optuna's representations. Categorical parameters are one-hot encoded.
+
+    .. seealso::
+        See an `example <https://github.com/optuna/optuna/blob/master/examples/multi_objective/
+        botorch_simple.py>`_ how to use the sampler.
+
+    .. seealso::
+        See the `BoTorch <https://botorch.org/>`_ homepage for details and for how to implement
+        your own ``candidates_func``.
+
+    .. note::
+        An instance of this sampler *should be not used with different studies* when used with
+        constraints. Instead, a new instance should be created for each new study. The reason for
+        this is that the sampler is stateful keeping all the computed constraints.
+
+    Args:
+        candidates_func:
+            An optional function that suggests the next candidates. It must take the training
+            data, the objectives, the constraints, the search space bounds and return the next
+            candidates. The arguments are of type ``torch.Tensor``. The return value must be a
+            ``torch.Tensor``. However, if ``constraints_func`` is omitted, constraints will be
+            :obj:`None`.
+
+            If omitted, is determined automatically based on the number of objectives and
+            constraints. If the number of objectives is one and no constraints are defined,
+            Upper Confidence Bound (UCB) is used. If the number of objectives is one and
+            constraints are defined, Expected Improvement (qEI) is used. If the number of
+            objectives are larger than one, Expected Hypervolume Improvement (qEHVI) is used.
+
+            The function should assume *maximization* of the objective.
+
+            .. seealso::
+                See :func:`optuna.integration.botorch.ucb_candidates_func` for an example.
+        constraints_func:
+            An optional function that computes the objective constraints. It must take a
+            :class:`~optuna.multi_objective.study.MultiObjectiveStudy`, a
+            :class:`~optuna.multi_objective.trial.FrozenMultiObjectiveTrial` and return the
+            constraints. The return value must be a sequence of :obj:`float` s. A value strictly
+            larger than 0 means that a constraints is violated. A value equal to or smaller than 0
+            is considered feasible.
+
+            If omitted, no constraints will be passed to ``candidates_func`` nor taken into
+            account during suggestion if ``candidates_func`` is omitted.
+
+            .. note::
+                ``constraints_func`` is called once per trial for each trial on each worker during
+                distributed optimization. Therefore, during distributed optimization, this
+                function should be deterministic to ensure that all workers hold the same values.
```

Comment:
We might want to introduce some sort of trial post-processing to allow this in the future. C.f. https://github.com/optuna/optuna/issues/1986